shell-here.el

```diff
@@ -141,20 +141,20 @@ root."
       ;; We need to `cd'
       (when (not (string= (shell-here-stripslash
                            (expand-file-name default-directory)) target))
+        (setq target (substring target (string-match "[^\:]+\/[^\:\*\\\<\>\|]+" target)))
 
         ;; Save any input on the command line; `comint-kill-input'
         ;; calls `kill-region', which we have flet with a function
         ;; that returns the region before deleting it.
         ;; The (insert (prog1 …)) inserts it (or an empty string, if
         ;; we know we have a new buffer) back into the shell buffer
         ;; after having changed directories.
-        (flet ((kill-region (start end)
+        (cl-flet ((kill-region (start end)
                  (prog1
                      (buffer-substring start end) (delete-region start end))))
           (insert (prog1 (or (when (not new) (comint-kill-input)) "")
                     (insert (concat "cd " (shell-quote-argument target)))
-                    (comint-send-input)
-                    (shell-cd target))))))))
```

Comment:
I think the `SHELL-CD` is still necessary for directory tracking to work -- at least that's my best recollection.  Do you know if that's changed?  This code is probably 15 years old.

Reply:
@ieure I don't think it's still necessary, and I couldn't find related stuff [here](https://www.gnu.org/software/emacs/manual/html_node/emacs/Directory-Tracking.html) as well... 

But I do know when that `(shell-cd target)` command is there and we're in tramp, it will give the error `cd: No such directory found via CDPATH environment variable` :)